link.go

```diff
@@ -71,41 +71,163 @@ type link struct {
+		Source:                   new(frames.Source),
+	}
+
+	if opts == nil {
+		return l, nil
 	}
 
-	// configure options
-	for _, o := range opts {
-		err := o(l)
-		if err != nil {
+	for _, v := range opts.Capabilities {
+		l.Source.Capabilities = append(l.Source.Capabilities, encoding.Symbol(v))
+	}
+	if opts.Durability > DurabilityUnsettledState {
+		return nil, fmt.Errorf("invalid Durability %d", opts.Durability)
+	}
+	l.Source.Durable = opts.Durability
+	if opts.DynamicAddress {
+		l.Target.Address = ""
+		l.dynamicAddr = opts.DynamicAddress
+	}
+	if opts.ExpiryPolicy != "" {
+		if err := encoding.ValidateExpiryPolicy(opts.ExpiryPolicy); err != nil {
 			return nil, err
 		}
+		l.Source.ExpiryPolicy = opts.ExpiryPolicy
+	}
+	l.Source.Timeout = opts.ExpiryTimeout
+	l.detachOnDispositionError = !opts.IgnoreDispositionErrors
+	if opts.Name != "" {
+		l.Key.name = opts.Name
+	}
+	if opts.Properties != nil {
+		l.properties = make(map[encoding.Symbol]interface{})
+		for k, v := range opts.Properties {
+			if k == "" {
+				return nil, errors.New("link property key must not be empty")
+			}
+			l.properties[encoding.Symbol(k)] = v
+		}
+	}
+	if opts.RequestedReceiverSettleMode != nil {
+		if rsm := *opts.RequestedReceiverSettleMode; rsm > ModeSecond {
+			return nil, fmt.Errorf("invalid RequestedReceiverSettleMode %d", rsm)
+		}
+		l.ReceiverSettleMode = opts.RequestedReceiverSettleMode
+	}
+	if opts.SettlementMode != nil {
+		if ssm := *opts.SettlementMode; ssm > ModeMixed {
+			return nil, fmt.Errorf("invalid SettlementMode %d", ssm)
+		}
+		l.SenderSettleMode = opts.SettlementMode
+	}
+	l.Source.Address = opts.SourceAddress
+	return l, nil
+}
+
+func newReceivingLink(source string, s *Session, r *Receiver, opts *ReceiverOptions) (*link, error) {
+	l := &link{
+		Key:                      linkKey{randString(40), encoding.RoleReceiver},
+		Session:                  s,
+		receiver:                 r,
+		close:                    make(chan struct{}),
+		Detached:                 make(chan struct{}),
+		ReceiverReady:            make(chan struct{}, 1),
+		detachOnDispositionError: true,
+		Source:                   &frames.Source{Address: source},
+		Target:                   new(frames.Target),
 	}
 
-	// sending unsettled messages when the receiver is in mode-second is currently
-	// broken and causes a hang after sending, so just disallow it for now.
-	if r == nil && senderSettleModeValue(l.SenderSettleMode) != ModeSettled && receiverSettleModeValue(l.ReceiverSettleMode) == ModeSecond {
-		return nil, errors.New("sender does not support exactly-once guarantee")
+	if opts == nil {
+		return l, nil
+	}
+
+	l.receiver.batching = opts.Batching
+	if opts.BatchMaxAge > 0 {
+		l.receiver.batchMaxAge = opts.BatchMaxAge
+	}
+	for _, v := range opts.Capabilities {
+		l.Target.Capabilities = append(l.Target.Capabilities, encoding.Symbol(v))
 	}
+	if opts.Credit > 0 {
+		l.receiver.maxCredit = opts.Credit
```

Comment:
Not in this PR, but it seems like Session.NewReceiver()'s code should just move into here instead. Now that you've split it up into separate new*Link's functions it'd probably be okay.